devtools: Add markdown dev panel.

Riken-Shah opened the pull request:
#15021

This PR adds a new developer tool: The "Markdown
Developer Panel" which will help us to instantly view:
- Rendered Output from marked.js (frontend-markdown).
- Rendered Output from Python Markdown (backend
markdown).
- Difference between the above two results.

Also added a feature to,
- Quickly test with different test cases from `markdown_test_cases.json`.


<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
Tested it in dev environment.


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


https://user-images.githubusercontent.com/63820270/115924385-45097300-a49d-11eb-9c0a-ab37b5eb2ad0.mp4







<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

Riken-Shah committed:
devtools: Add markdown dev panel.

This commit adds a new developer tool,
"Markdown Developer Panel" which will help us:
-> View markdown output from marked.js (frontend
markdown).
-> View markdown output from Python Markdown
(backend markdown).
-> View the difference between the above two results.

Fixes #15021.

Riken-Shah commented on static/js/portico/markdown_panel.js:
```diff
@@ -0,0 +1,208 @@
+
+import * as channel from "../channel";
+import * as markdown from "../markdown";
+// Importing stream_data to avoid import error
```
While importing `get_helpers` from `markdown_config.js`.
I am facing the following error,
```bash
Uncaught TypeError: _filter__WEBPACK_IMPORTED_MODULE_25__.Filter is not a constructor
    at new MessageListData (app~archive~markdown-panel.js:20365)
    at Module../static/js/all_messages_data.js (app~archive~markdown-panel.js:1436)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/stream_topic_history.js (app~archive~markdown-panel.js:51712)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/stream_data.js (app~archive~markdown-panel.js:47430)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/filter.js (app~archive~markdown-panel.js:11094)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/narrow_state.js (app~archive~markdown-panel.js:25623)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/hash_util.js (app~archive~markdown-panel.js:13270)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/markdown_config.js (markdown-panel.js:1627)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/portico/markdown_panel.js (markdown-panel.js:1764)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Object.41 (markdown-panel.js:2008)
    at __webpack_require__ (markdown-panel.js:854)
    at checkDeferredModules (markdown-panel.js:46)
    at markdown-panel.js:1590
    at markdown-panel.js:1593
```

I am currently importing `stream_data.js` to avoid this.

Can someone help me with this?

Riken-Shah commented on static/js/portico/markdown_panel.js:
```diff
@@ -0,0 +1,208 @@
+// Importing stream_data to avoid import error
+// while importing markdown_config.
+import * as stream_data from "../stream_data"; // eslint-disable-line no-unused-vars
+import {get_helpers} from "../markdown_config"; // eslint-disable-line import/order
```
While importing `get_helpers` from `markdown_config.js`.
I am facing the following error,
```bash
Uncaught TypeError: _filter__WEBPACK_IMPORTED_MODULE_25__.Filter is not a constructor
    at new MessageListData (app~archive~markdown-panel.js:20365)
    at Module../static/js/all_messages_data.js (app~archive~markdown-panel.js:1436)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/stream_topic_history.js (app~archive~markdown-panel.js:51712)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/stream_data.js (app~archive~markdown-panel.js:47430)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/filter.js (app~archive~markdown-panel.js:11094)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/narrow_state.js (app~archive~markdown-panel.js:25623)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/hash_util.js (app~archive~markdown-panel.js:13270)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/markdown_config.js (markdown-panel.js:1627)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Module../static/js/portico/markdown_panel.js (markdown-panel.js:1764)
    at __webpack_require__ (markdown-panel.js:854)
    at fn (markdown-panel.js:151)
    at Object.41 (markdown-panel.js:2008)
    at __webpack_require__ (markdown-panel.js:854)
    at checkDeferredModules (markdown-panel.js:46)
    at markdown-panel.js:1590
    at markdown-panel.js:1593
```

I am currently importing `stream_data.js` to avoid this.

Can someone help me with this?

ganpa3 commented:
This PR and #18102 are exactly the same (except #18102 having fixtures)
@akshatdalton FYI

ganpa3 reviewed (commented):
This PR and #18102 are exactly the same (except #18102 having fixtures)
@akshatdalton FYI

Riken-Shah commented:
> This PR and #18102 are exactly the same (except #18102 having fixtures)
> @akshatdalton FYI

@ganpa3 Actually I and @akshatdalton  are collaborating on this issue. 
I am working on the dev panel part and @akshatdalton is working on the fixture part.

ganpa3 commented:
@Riken-Shah Well, I do know that. Wouldn't it better to have both in a single commit, as we do in integrations, to have a more complete commit, maybe [co-authoring](https://docs.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors)?

ganpa3 commented:
Suggestion: Pretty printing the output would make it more readable

Riken-Shah commented:
@ganpa3 Thanks, I guess this 

> Suggestion: Pretty printing the output would make it more readable

Yep, we are already doing it with the help of `prettier`.

akshatdalton committed:
devtools: Add support for markdown test fixtures.

This commit adds support for markdown test fixtures in
`/devtools/markdown`.

A dropdown is added in `/devtools/markdown` which is populated
with name of `regular_tests` from
`zerver/tests/fixtures/markdown_test_cases.json`.

Tests are added for the new API: `devtools/markdown/<fixture_name>/fixture`
in `test_markdown_dev_panel.py`.

Riken-Shah commented:
@amanagr Can you please review this?

Also, Do you think should I split the first commit?

timabbott commented on package.json:
```diff
@@ -45,6 +46,7 @@
+    "highlight.js": "^10.7.2",
     "html-webpack-plugin": "^4.0.0-beta.8",
     "jquery": "^3.4.1",
     "jquery-caret-plugin": "^1.5.2",
```
Can you move the new dependencies to an independent first commit?  That way we can merge quickly the main thing that might merge conflict.

Also, do these belong in `devDependencies`, given that this is a page only used in a development environment?  

(It's possible that `webpack` considerations limit our ability to do that).

timabbott commented on static/js/portico/markdown_dev_panel.js:
```diff
@@ -0,0 +1,241 @@
+            data: {content: message},
+            beforeSend(xhr) {
+                xhr.setRequestHeader("X-CSRFToken", $("#csrf_token").val());
+            },
```
I would have expected this to be done by importing `static/js/csrf.js`.

timabbott commented on static/styles/portico/markdown_dev_panel.css:
```diff
@@ -0,0 +1,140 @@
+button {
+    background-color: hsl(152, 52%, 91%);
+    color: hsl(146, 50%, 36%);
+}
```
We should write these CSS rules to be contained to only apply to this page.

Even if it's in its own bundle and thus can only impact that page, it's not a good coding pattern.

Could all the CSS in this file by inside `#markdown-panel`?

Riken-Shah replied:
I rearranged CSS rules in their respective parent selector. Now most of the rules are inside `#markdown-panel` but some rules for navbar and spacing needs to be outside, 
```css
.top-navbar {
    display: grid;
    grid-template-columns: 1fr 5fr 1fr;
}

.centerpiece {
    font-size: 30px;
    text-align: center;
}

.pad-small {
    height: 100px;
    width: 100%;
}

.pad-tiny {
    height: 50px;
    width: 100%;
}
```
P.S: There is one fix for this - We can wrap `portico_content` in `#markdown_dev_panel_page` and then wrap all css rules inside `#markdown_dev_panel_page`.

timabbott commented on tools/linter_lib/custom_check.py:
```diff
@@ -317,8 +317,8 @@
-            "exclude": {"zerver/tests"},
-            "description": "Argument to json_error should a literal string enclosed by _()",
+            "exclude": {"zerver/tests", "zerver/views/development/"},
+            "description": "Argument to json_error should be a literal string enclosed by _()",
```
I merged this commit as 3da6f3c0afe92d0d5e5584076518a482fcb6a5a8, after a small commit message tweak.

Riken-Shah committed:
dependencies: Add `diff2html` and `highlight.js` to the list.

This is a prep commit to add markdown dev pannel.

timabbott commented:
I merged f2c7e8e80fadd2064901873c1e935219a1bfe1f3 and rebased this.  It doesn't run -- there's a few codebase migrations that need to be applied.  Feel free to fix at your convenience; this isn't urgent but I would like to integrate it.

timabbott commented:
I still would like this to get finished, added the completion candidate label.
